File: networking/forwarder/lib/ipc/pipe_duplex.go
//go:build unix

package ipc

import (
	"bytes"
	"context"
	"errors"
	"io/fs"
	"lib"
	"log"
	"os"
	"sync"
	"syscall"
	"time"

	"github.com/pdgendt/cobs"
	"golang.org/x/sync/errgroup"
	"golang.org/x/sys/unix"
)

var (
	ErrInOutPipesAreSame   = errors.New("the in-pipe and out-pipe are the same")
	ErrExistingFileNotFifo = errors.New("the existing file is not a FIFO")
)

const (
	pipeDuplexOpenReaderFlags = syscall.O_RDONLY | syscall.O_NONBLOCK
	pipeDuplexOpenWriterFlags = syscall.O_WRONLY | syscall.O_NONBLOCK
	pipeDuplexModeFlags       = syscall.S_IRUSR | syscall.S_IWUSR | syscall.S_IRGRP | syscall.S_IROTH
	pipeDuplexPollInterval    = 50 * time.Millisecond
	pipeDuplex_PIPE_BUF       = 4096
)

// Signal messages range from 1 to 255 & indicate control flow for the bytestream of the pipe.
type SignalMessage byte

const (
	// DISCARD_PREVIOUS tells the receiver to discard previous partial work.
	DiscardPrevious SignalMessage = 0x01
)

type OnMessage = func(msg []byte) error

// Creates a named-pipe communication duplex. Creates a named-pipe communication duplex.
// The reader end is responsible for creating the pipe.
//
// The layers are:
//  1. Raw binary data over pipes
//  2. Variable-length binary packets with COBS
//  3. JSON-like values with Message Pack
type PipeDuplex struct {
	inPath  string
	outPath string

	rawOutMu sync.Mutex
	rawOut   chan []byte

	ctx    context.Context
	cancel context.CancelFunc
	errg   *errgroup.Group
}

func NewPipeDuplex(inPath, outPath string, onMessage OnMessage) (*PipeDuplex, error) {
	// they must be different files
	if inPath == outPath {
		return nil, ErrInOutPipesAreSame
	}
	// pipes should only ever be created, and only by the reader (one-way operations)
	if err := ensureFifoExists(inPath); err != nil {
		return nil, err
	}

	ctx, cancel := context.WithCancel(context.Background())
	errg, ctx := errgroup.WithContext(ctx)
	p := &PipeDuplex{
		inPath:  inPath,
		outPath: outPath,

		rawOut: make(chan []byte, 128), // TODO: decide on size of this w/ constant??

		ctx:    ctx,
		cancel: cancel,
		errg:   errg,
	}
	// Reader
	p.errg.Go(func() error {
		return p.pipeBufferReader(onMessage)
	})

	// Writer
	p.errg.Go(func() error {
		return p.pipeBufferWriter()
	})

	return p, nil
}

// Close stops all goroutines and waits for them to exit.
func (p *PipeDuplex) Close() error {
	p.cancel()

	// this channel is exclusively written to via methods on this object handle, so it is its owner;
	// owners must be the ones to close channels to avoid race conditions
	defer func() {
		// lock channel to avoid race conditions when closing
		p.rawOutMu.Lock()
		defer p.rawOutMu.Unlock()

		close(p.rawOut)
	}()

	return p.errg.Wait()
}

// SendMessage MessagePack-encodes a "value" and enqueues it to the writer.
func (p *PipeDuplex) SendMessage(msg []byte) error {
	// lock channel to avoid race conditions when closing
	p.rawOutMu.Lock()
	defer p.rawOutMu.Unlock()

	// send message bytes over outRaw channel
	select {
	case p.rawOut <- msg:
		// TODO: could this trigger a race condition if calling Close() immediately after SendMessage()???
		//  should I lock p.rawOut w/ a mutex??
		return nil
	case <-p.ctx.Done():
		return nil
	}
}

func (p *PipeDuplex) InPath() string  { return p.inPath }
func (p *PipeDuplex) OutPath() string { return p.outPath }

// ===== Private =====

func ensureFifoExists(path string) error {
	// try to make a file if one doesn't exist already
	// TODO: add equivalent of `ensure_parent_directory_exists(path)` here !!!!!! <- may cause bugs w/out it???
	if err := unix.Mkfifo(path, pipeDuplexModeFlags); err != nil {
		if errno, ok := err.(unix.Errno); ok {
			// misc error, do not handle
			if errno != unix.EEXIST {
				return err
			}

			// ensure the file exists is FIFO
			fi, err := os.Stat(path)
			if err != nil {
				return err // misc error, do not handle
			}
			if fi.Mode()&fs.ModeNamedPipe == 0 {
				return ErrExistingFileNotFifo
			}
			return nil
		} else {
			return err // misc error, do not handle
		}
	}
	return nil
}

func (p *PipeDuplex) pipeBufferReader(onMessage OnMessage) error {
	// open reader in nonblocking mode -> should not fail & immediately open;
	// this marks when the writer process has "started"
	fd, err := unix.Open(p.inPath, pipeDuplexOpenReaderFlags, pipeDuplexModeFlags)
	if err != nil {
		return err
	}
	defer unix.Close(fd)

	// continually pull from the pipe and interpret messages as such:
	//  - all messages are separated/framed by NULL bytes (zero)
	//  - messages with >=2 bytes are COBS-encoded messages, because
	//    the smallest COBS-encoded message is 2 bytes
	//  - 1-byte messages are therefore to be treated as control signals
	var buf []byte // accumulation buffer
	for {
		select { // check for kill-signal
		case <-p.ctx.Done():
			return nil
		default:
		}

		// read available data (and try again if nothing)
		data := make([]byte, pipeDuplex_PIPE_BUF)
		n, err := unix.Read(fd, data)
		if err != nil {
			errno, ok := err.(unix.Errno)
			if !ok || errno != unix.EAGAIN {
				return err
			}

			// if there is a writer connected & the buffer is empty, this would block
			// so we must consume this error gracefully and try again
			time.Sleep(pipeDuplexPollInterval)
			continue
		}
		if n == 0 {
			time.Sleep(pipeDuplexPollInterval)
			continue
		}

		// extend buffer with new data
		buf = append(buf, data[:n]...)

		// if there are no NULL bytes in the buffer, no new message has been formed
		chunks := bytes.Split(buf, []byte{0x00})
		if len(chunks) == 1 {
			continue
		}

		// last chunk is always an unfinished message, so that becomes our new buffer;
		// the rest should be decoded as either signals or COBS and put on queue
		buf = chunks[len(chunks)-1]
		for i := 0; i < len(chunks)-1; i++ {
			chunk := chunks[i]

			// ignore empty messages (they mean nothing)
			if len(chunk) == 0 {
				continue
			}

			// interpret 1-byte messages as signals (they indicate control-flow on messages)
			if len(chunk) == 1 {
				log.Printf("(reader): gotten control signal: %v", chunk[0])
				// TODO: do some kind of stuff here??
				continue
			}

			// interpret >=2 byte messages as COBS-encoded data (decode them)
			decoded, err := cobs.Decode(chunk)
			if err != nil {
				return err
			}

			// call the callback to handle message
			if err := onMessage(decoded); err != nil {
				return err
			}
		}
	}
}

func (p *PipeDuplex) pipeBufferWriter() error {
	log.Printf("(writer): started")

	// continually attempt to open FIFO for reading in nonblocking mode -> will error that:
	//  - ENOENT[2] No such file or directory: until a reader creates FIFO
	//  - ENXIO[6] No such device or address: until a reader opens FIFO
	fd := -1
	for {
		select { // check for kill-signal
		case <-p.ctx.Done():
			return nil
		default:
		}

		tempFd, err := unix.Open(p.outPath, pipeDuplexOpenWriterFlags, pipeDuplexModeFlags)
		if err != nil {
			if errno, ok := err.(unix.Errno); ok {
				// misc error, do not handle
				if !(errno == unix.ENOENT || errno == unix.ENXIO) {
					return err
				}

				// try again if waiting for FIFO creation or reader-end opening
				time.Sleep(pipeDuplexPollInterval)
				continue
			} else {
				return err // misc error, do not handle
			}
		}
		fd = tempFd
		defer unix.Close(fd)

		// ensure the file exists is FIFO
		mode, err := lib.FstatGetMode(fd)
		if err != nil {
			return err // misc error, do not handle
		}
		if mode&fs.ModeNamedPipe == 0 {
			return ErrExistingFileNotFifo
		}

		break // continue logic
	}

	// read bytes from rawOut & write them to pipe
	for {
		select {
		case buf, ok := <-p.rawOut:
			if !ok {
				return nil
			}
			if err := p.writeData(fd, buf); err != nil {
				return err
			}
		case <-p.ctx.Done():
			return nil
		}

	}
}

func (p *PipeDuplex) writeData(fd int, buf []byte) error {
	// COBS-encode the data & append NULL-byte to signify end-of-frame
	buf, err := cobs.Encode(buf)
	if err != nil {
		return err
	}
	buf = append(buf, 0x00)
	total := len(buf)
	sent := 0

	// begin transmission progress
	for sent < total {
		select { // check for kill-signal
		case <-p.ctx.Done():
			return nil
		default:
		}

		// write & progress on happy path
		written, err := unix.Write(fd, buf[sent:])
		if err == nil {
			sent += written
			continue
		}

		// cast to OS error for propper handling
		errno, ok := err.(unix.Errno)
		if !ok {
			return err // misc error, do not handle
		}

		// non-blocking pipe is full, wait a bit and retry
		if errno == syscall.EAGAIN {
			time.Sleep(pipeDuplexPollInterval)
			continue
		}

		// reader disconnected -> handle failure-recovery by doing:
		//  1. signal DISCARD_PREVIOUS to any reader
		//  2. re-setting the progress & trying again
		if errno == syscall.EPIPE {
			if err := p.writeSignal(fd, DiscardPrevious); err != nil {
				return err
			}
			sent = 0
			continue
		}

		return err // misc error, do not handle
	}
	return nil
}

func (p *PipeDuplex) writeSignal(fd int, sig SignalMessage) error {
	signalMessageLength := 2

	// Turn signal-byte into message by terminating with NULL-byte
	buf := []byte{byte(sig), 0x00}
	lib.Assert(len(buf) == signalMessageLength, "this must never NOT be the case")

	// attempt to write until successful
	for {
		select { // check for kill-signal
		case <-p.ctx.Done():
			return nil
		default:
		}

		// small writes (e.g. 2 bytes) should be atomic as per Pipe semantics,
		// meaning IF SUCCESSFUL: the number of bytes written MUST be exactly 2
		written, err := unix.Write(fd, buf)
		if err == nil {
			lib.Assert(written == signalMessageLength, "this must never NOT be the case")
			break
		}

		// cast to OS error for propper handling
		errno, ok := err.(unix.Errno)
		if !ok {
			return err // misc error, do not handle
		}

		// wait a bit and retry if:
		//  - non-blocking pipe is full
		//  - the pipe is broken because of reader disconnection
		if errno == syscall.EAGAIN || errno == syscall.EPIPE {
			time.Sleep(pipeDuplexPollInterval)
			continue
		}

		return err // misc error, do not handle
	}
	return nil
}

File: src/exo/shared/ipc/pipe_duplex.py
"""
SEE:
 - https://pubs.opengroup.org/onlinepubs/007904875/functions/open.html
 - https://man7.org/linux/man-pages/man2/openat.2.html
 - https://man7.org/linux/man-pages/man3/mkfifo.3.html
 - https://man7.org/linux/man-pages/man7/pipe.7.html

TODO: add locking on reader/writer ends to prevent multiwriters??
TODO: use signal bytes to ensure proper packet consistency
      +stretch: implement packet IDs, retries, dual-stream confirmations, RPCs & so on

TODO: for more hardening -> check if any of the syscalls used return signal interrupt errors (like in the locking case)
      and interrupt on that happening -> this may not be an issue PER SE but might potentially create insanely bizzare bugs
      if it happens that this behavior DOES occasionally happen for no apparent reason

TODO: maybe consider padding all messages with 0s on both ends ?? so as to prevent ANY ambiguous boundaries ever!!
"""

import errno
import logging
import multiprocessing
import os
import queue
import stat
import threading
import time
from enum import Enum
from multiprocessing.queues import Queue as MQueueT
from multiprocessing.synchronize import Event as MEventT
from threading import Event as TEventT
from typing import Callable

from cobs import cobs  # pyright: ignore[reportMissingTypeStubs]
from pytest import LogCaptureFixture

from exo.shared.utils.fs import (
    StrPath,
    delete_if_exists,
    ensure_parent_directory_exists,
)

OPEN_READER_FLAGS = os.O_RDONLY | os.O_NONBLOCK
OPEN_WRITER_FLAGS = os.O_WRONLY | os.O_NONBLOCK

# 0x644 mode flags -> user has read-write permissions, others have read permission only
MODE_FLAGS = stat.S_IRUSR | stat.S_IWUSR | stat.S_IRGRP | stat.S_IROTH

POLL_INTERVAL = 0.05  # TODO: maybe parametrize this in classes??
PIPE_BUF = 4096  # size of atomic writes on (most) UNIX pipes


class SignalMessage(Enum):
    """
    Signal messages range from 1 to 255 & indicate control flow for the bytestream of the pipe.

    """

    DISCARD_PREVIOUS = b"\x01"


class PipeDuplex:
    """
    Creates a named-pipe communication duplex. The reader end is responsible for creating the pipe.

    The layers are:
      1. Raw binary data over pipes
      2. Variable-length binary packets with COBS
      3. JSON-like values with Message Pack
    """

    def __init__(
        self,
        in_pipe: StrPath,
        out_pipe: StrPath,
        in_callback: Callable[[bytes], None],
    ):
        assert in_pipe != out_pipe  # they must be different files

        # pipes should only ever be created, and only by the reader (one-way operations)
        _ensure_fifo_exists(in_pipe)  # ensures reader pipe exists

        # create readonly properties (useful for inspection)
        self._in_pipe = in_pipe
        self._out_pipe = out_pipe

        # init synchronisation variables
        self._mkill = multiprocessing.Event()
        self._tkill = threading.Event()
        in_mq: MQueueT[bytes] = multiprocessing.Queue()
        self._out_mq: MQueueT[bytes] = multiprocessing.Queue()
        in_mstarted = multiprocessing.Event()

        # process for reading in binary messages from pipe
        self._p_in = multiprocessing.Process(
            target=_pipe_buffer_reader,
            args=(in_pipe, in_mq, in_mstarted, self._mkill),
            daemon=True,
        )
        self._p_in.start()

        # thread for pulling down binary messages from message queue & calling the callback
        self._t_in = threading.Thread(
            target=_binary_object_dispatcher,
            args=(in_mq, in_callback, self._tkill),
            daemon=True,
        )
        self._t_in.start()

        # process to write binary messages to pipe
        out_mstarted = multiprocessing.Event()
        self._p_out = multiprocessing.Process(
            target=_pipe_buffer_writer,
            args=(out_pipe, self._out_mq, out_mstarted, self._mkill),
            daemon=True,
        )
        self._p_out.start()

        # wait for processes to start properly
        in_mstarted.wait()
        out_mstarted.wait()

    def __del__(self):
        # signal to these processes to die (if they haven't already)
        self._mkill.set()
        self._tkill.set()

    def send_message(self, msg: bytes):
        self._out_mq.put_nowait(msg)

    @property
    def in_pipe(self):
        return self._in_pipe

    @property
    def out_pipe(self):
        return self._out_pipe


def _ensure_fifo_exists(path: StrPath):
    # try to make a file if one doesn't exist already
    ensure_parent_directory_exists(path)
    try:
        os.mkfifo(path, mode=MODE_FLAGS)
    except OSError as e:
        # misc error, do not handle
        if e.errno != errno.EEXIST:
            raise

        # ensure the file exists is FIFO
        st = os.stat(path)
        if stat.S_ISFIFO(st.st_mode):
            return

        # this file is not FIFO
        raise FileExistsError(f"The file '{path}' isn't a FIFO") from e


def _pipe_buffer_reader(
    path: StrPath, mq: MQueueT[bytes], started: MEventT, kill: MEventT
):
    # TODO: right now the `kill` control flow is somewhat haphazard -> ensure every loop-y or blocking part always
    #       checks for kill.is_set() and returns/cleans up early if so

    # open reader in nonblocking mode -> should not fail & immediately open;
    # this marks when the writer process has "started"
    fd = os.open(path, OPEN_READER_FLAGS)
    started.set()
    print("(reader):", "started")

    # continually pull from the pipe and interpret messages as such:
    #  - all messages are separated/framed by NULL bytes (zero)
    #  - messages with >=2 bytes are COBS-encoded messages, because
    #    the smallest COBS-encoded message is 2 bytes
    #  - 1-byte messages are therefore to be treated as control signals
    #
    # TODO: right now i just need to get this to work, but the scheme is fundamentally
    #       extensible for robustness, e.g. signal-bytes can be used to drive state-machines
    #       for ensuring message atomicity/transmission
    #       e.g. we can use single-bytes to discriminate COBS values to say "this is length of upcoming message"
    #            vs. this is the actual content of the message, and so on
    #       .
    #       BUT for now we can just use signal (0xff 0x00) to mean "discard previous message" or similar...
    #       .
    #       BUT in THEORY we very well could have something like
    #       (0x10 0x00)[header signal] + (...)[header data like length & so on]
    #       + (0x20 0x00)[body signal] + (...)[body data]
    #       + (0x30 0x00)[checksum signal] + (...)[checksum data]
    #       And requests to re-send messages that were lost, and so on, like this is a fully 2-layer duplex
    #       communication so we could turn this into a VERY powerful thing some time in the future, like
    #       a whole-ass reimplementation of TCP/PIPES lmaooooo
    buffer = bytearray()
    while not kill.is_set():
        try:
            # read available data (and try again if nothing)
            try:
                data = os.read(fd, PIPE_BUF)
                if data == b"":
                    time.sleep(POLL_INTERVAL)
                    continue
            except OSError as e:
                if e.errno != errno.EAGAIN:
                    raise

                # if there is a writer connected & the buffer is empty, this would block
                # so we must consume this error gracefully and try again
                time.sleep(POLL_INTERVAL)
                continue

            # extend buffer with new data
            buffer.extend(data)

            # if there are no NULL bytes in the buffer, no new message has been formed
            chunks = buffer.split(sep=b"\x00")
            if len(chunks) == 1:
                continue

            # last chunk is always an unfinished message, so that becomes our new buffer;
            # the rest should be decoded as either signals or COBS and put on queue
            buffer = chunks.pop()
            for chunk in chunks:
                chunk = bytes(chunk)

                # ignore empty messages (they mean nothing)
                if chunk == b"":
                    continue

                # interpret 1-byte messages as signals (they indicate control-flow on messages)
                if len(chunk) == 1:
                    print("(reader):", f"gotten control signal: {chunk[0]}")
                    continue  # TODO: right now they should be ignored, since I'm not sure what I want them to do

                # interpret >=2 byte messages as COBS-encoded data (decode them)
                decoded = cobs.decode(chunk)  # pyright: ignore[reportUnknownMemberType]
                mq.put(decoded)
        except BaseException as e:
            # perform cleanup & log before re-raising
            os.close(fd)
            logging.error(msg=f"Error when reading from named pipe at '{path}': {e}")
            raise
    os.close(fd)


def _binary_object_dispatcher(
    mq: MQueueT[bytes], callback: Callable[[bytes], None], kill: TEventT
):
    while not kill.is_set():
        # try to get with timeout (to allow to read the kill-flag)
        try:
            message = mq.get(block=True, timeout=POLL_INTERVAL)
        except queue.Empty:
            continue

        # dispatch binary object with callback
        callback(message)


def _pipe_buffer_writer(
    path: StrPath, mq: MQueueT[bytes], started: MEventT, kill: MEventT
):
    # TODO: right now the `kill` control flow is somewhat haphazard -> ensure every loop-y or blocking part always
    #       checks for kill.is_set() and returns/cleans up early if so

    # for now, started events for writer are rather vacuous: TODO: remove or make more usefull??
    started.set()
    print("(writer):", "started")

    # continually attempt to open FIFO for reading in nonblocking mode -> will error that:
    #  - ENOENT[2] No such file or directory: until a reader creates FIFO
    #  - ENXIO[6] No such device or address: until a reader opens FIFO
    fd = None
    while not kill.is_set():
        try:
            fd = os.open(path, os.O_WRONLY | os.O_NONBLOCK)

            # ensure the file exists is FIFO
            st = os.fstat(fd)
            print("mode:", st.st_mode & 0o170000)
            if stat.S_ISFIFO(st.st_mode):
                break

            # cleanup on error
            os.close(fd)
            raise FileExistsError(f"The file '{path}' isn't a FIFO")
        except FileExistsError:
            raise  # propagate error
        except OSError as e:
            # misc error, do not handle
            if not (e.errno == errno.ENOENT or e.errno == errno.ENXIO):
                raise

            # try again if waiting for FIFO creation or reader-end opening
            time.sleep(POLL_INTERVAL)
            continue
    assert fd is not None

    while not kill.is_set():
        try:
            # try to get with timeout (to allow to read the kill-flag)
            try:
                data = mq.get(block=True, timeout=POLL_INTERVAL)
            except queue.Empty:
                continue

            # write all data (by continually re-trying until it is done)
            _write_data(fd, data)
        except BaseException as e:
            # perform cleanup & log before re-raising
            os.close(fd)
            logging.error(msg=f"Error when writing to named pipe at '{path}': {e}")
            raise

    os.close(fd)


def _write_data(fd: int, buf: bytes):
    # COBS-encode the data & append NULL-byte to signify end-of-frame
    buf = cobs.encode(buf) + b"\x00"  # pyright: ignore[reportUnknownMemberType]
    total = len(buf)
    sent = 0

    # begin transmission progress
    while sent < total:
        try:
            # Write remaining bytes to the pipe
            written = os.write(fd, buf[sent:])
            sent += written
        except OSError as e:
            # non-blocking pipe is full, wait a bit and retry
            if e.errno == errno.EAGAIN:
                time.sleep(POLL_INTERVAL)
                continue

            # reader disconnected -> handle failure-recovery by doing:
            #  1. signal DISCARD_PREVIOUS to any reader
            #  2. re-setting the progress & trying again
            if e.errno == errno.EPIPE:
                _write_signal(fd, SignalMessage.DISCARD_PREVIOUS)
                sent = 0
                continue

            raise  # misc error, do not handle


def _write_signal(fd: int, signal: SignalMessage):
    signal_message_length = 2

    # Turn signal-byte into message by terminating with NULL-byte
    buf = signal.value + b"\x00"
    assert len(buf) == signal_message_length

    # attempt to write until successful
    while True:
        try:
            # small writes (e.g. 2 bytes) should be atomic as per Pipe semantics,
            # meaning IF SUCCESSFUL: the number of bytes written MUST be exactly 2
            written = os.write(fd, buf)
            assert written == signal_message_length
            break
        except OSError as e:
            # wait a bit and retry if:
            #  - non-blocking pipe is full
            #  - the pipe is broken because of reader disconnection
            if e.errno == errno.EAGAIN or e.errno == errno.EPIPE:
                time.sleep(POLL_INTERVAL)
                continue

            raise  # misc error, do not handle


def _test_one_two_three():
    one_path = "/tmp/one.pipe"
    two_path = "/tmp/two.pipe"
    delete_if_exists(one_path)
    delete_if_exists(two_path)

    owner = PipeDuplex(
        in_pipe=one_path,
        out_pipe=two_path,
        in_callback=lambda x: print(f"wow, owner got: [{len(x)}]{x}"),
    )

    guest = PipeDuplex(
        in_pipe=two_path,
        out_pipe=one_path,
        in_callback=lambda x: print(f"wow, guest1 got: [{len(x)}]{x}"),
    )

    owner.send_message(bytes(0 for _ in range(10)))

    guest.send_message(bytes(0 for _ in range(200)))

    time.sleep(1)

    del guest
    guest = PipeDuplex(
        in_pipe=two_path,
        out_pipe=one_path,
        in_callback=lambda x: print(f"wow, guest2 got: [{len(x)}]{x}"),
    )

    guest.send_message(bytes(0 for _ in range(21)))

    owner.send_message(bytes(0 for _ in range(12)))

    time.sleep(1)

    delete_if_exists(one_path)
    delete_if_exists(two_path)


def test_running_pipe_duplex(caplog: LogCaptureFixture):
    caplog.set_level(logging.INFO)

    _test_one_two_three()
    time.sleep(1)
